### PR TITLE
Datatype handling when in write_raw_bids

### PR DESCRIPTION
**Pull request by richardkoehler**
PR Description
--------------

Related to #752 

The user must now specify the datatype when calling write_raw_bids. Not automatic datatype inferring is done anymore in write_raw_bids. However, it is checked if the specified datatype is present in the dataset.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: mne_bids/utils.py [hunk 1]
```diff
@@ -91,20 +91,29 @@ def _get_ch_type_mapping(fro='mne', to='bids'):
...
+    datatypes = list()
+    ieeg_types = ['seeg', 'ecog']
+    if any(ieeg_type in raw for ieeg_type in ieeg_types):
+        datatypes.append('ieeg')
+    if 'meg' in raw:
+        datatypes.append('meg')
+    if 'eeg' in raw:
+        datatypes.append('eeg')
+    if len(datatypes) == 0:
+        raise ValueError('No MEG, EEG or iEEG channels found in data.'
```
Comment (agramfort) [inline mne_bids/utils.py:103]:
```suggestion
        raise ValueError('No MEG, EEG or iEEG channels found in data. '
```

File: mne_bids/utils.py [hunk 2]
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
...
             timedelta(0, stamp[0], stamp[1]))  # day, sec, μs
+
+
+def _check_datatype(raw, datatype):
+    """Check if datatype exists in given raw object."""
+    supported_types = ['meg', 'eeg', 'ieeg']
+    if datatype not in supported_types:
+        raise ValueError(
+            f'The specified datatype {datatype} is currently not supported. '
+            'It should be one of  either `meg`, `eeg` or `ieeg`. Please '
```
Comment (agramfort) [inline mne_bids/utils.py:405]:
```suggestion
            f'It should be one of  either `meg`, `eeg` or `ieeg` (Got `{datatype}`}. Please '
```

File: mne_bids/utils.py [hunk 3]
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
...
+            'specify a valid datatype using '
+            '`bids_path.update(datatype="<datatype>")`.')
+    datatype_matches = False
+    if datatype == 'eeg':
+        if datatype in raw:
+            datatype_matches = True
```
Comment (agramfort) [inline mne_bids/utils.py:411]:
```suggestion
    if datatype == 'eeg' and datatype in raw:
        datatype_matches = True
```

File: mne_bids/utils.py [hunk 4]
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
...
+    elif datatype == 'meg':
+        if datatype not in raw:
```
Comment (agramfort) [inline mne_bids/utils.py:413]: idem

File: mne_bids/write.py [hunk 5]
```diff
@@ -1185,7 +1185,11 @@ def write_raw_bids(raw, bids_path, events_data=None,
                 'https://github.com/mne-tools/mne-bids/issues')
         raise ValueError(msg)
 
-    datatype = _handle_datatype(raw)
+    if bids_path.datatype is None:
+        datatype = _handle_datatype(raw)
+    else:
+        datatype = bids_path.datatype
+        _check_datatype(raw, datatype)
```
Comment (agramfort) [inline mne_bids/write.py:1192]: I would do this in the _handle_datatype function

`datatype = _handle_datatype(raw, bids_path.datatype)`

then you call _check_datatype in _handle_datatype and your return directly the 2nd parameter if valid.

**Review by agramfort — COMMENTED** [review]
just a few suggestions.

thx @richardkoehler

File: mne_bids/utils.py (hunk 4 above)
Comment (richardkoehler) [inline mne_bids/utils.py:413]:
```suggestion
    elif datatype == 'meg' and datatype in raw:
```

File: mne_bids/write.py (hunk 5 above)
Comment (richardkoehler) [inline mne_bids/write.py:1192]:
```suggestion
    datatype = _handle_datatype(raw)
```

**Comment by richardkoehler** [comment]
@agramfort Thanks for the helpful suggestions! Tests are currently failing because write_raw_bids() now makes no more automatic assumptions about the datatype if there are multiple ephys data types (meg, eeg or ieeg) in our Raw object and because it actually checks if the bids_path.datatype is present in the Raw object. I can start fixing the tests, but as I already wrote in #752, I think the current PR might "break" write_raw_bids() for a lot of existing code out there (as demonstrated by the failing tests). So I think we need to make sure that everyone agrees with the new behavior of write_raw_bids:

1. If bids_path.datatype is not set, _handle_datatype will try to infer the datatype:
    a) If there is a single data type (of meg, eeg, ieeg) in the Raw object, this type will be picked
    b) If there are multiple types (of meg, eeg, ieeg) in the Raw object, an Error will be thrown, asking the user to specify the  data type
    c)  If no data type (of meg, eeg, ieeg) is found in the Raw object, an Error will be thrown, asking the user to specify the channel types and/or bids_path.datatype
2. If bids_path.datatype is set, _handle_datatype will call _check_datatype and verify that the specified data type is actually present in the Raw object. If it is not, an Error will be thrown.

Alternatively, we could "trust" the user to know what he is doing, and simply issue a warning message for case 2. There might be some cases where a user might want to write the data without specifying the channel types, even though I can't currently think of any. Is it really the job of MNE-BIDS to enforce these kind of things?
@jasmainak

**Comment by agramfort** [comment]
I would say that if you have no datatype and the data contains MEG and EEG
then you should
infer automatically it's an meg data.

would this already avoid breaking many tests?

**Comment by richardkoehler** [comment]
@agramfort  Yes, this actually fixed many of the tests. I implemented this way of handling datatype=None and fixed the remaining failing tests. In the case of datatype='eeg' with additional 'meg' channels in the data we are now trusting the user to know what he is doing, and NOT enforcing the datatype 'meg'. Therefore, the user can decide to write out combined MEG-EEG recordings in an EEG data format.

**Comment by richardkoehler** [comment]
@agramfort  Actually, it won't be possible as I get the following Error:  
```
E           UserWarning: Encountered unsupported non-voltage units: T/m, T
E           Note that the BrainVision format specification supports only µV.
```
So apparently it doesn't make any sense to save MEG-EEG data as EEG data because brainvision doesn't support MEG units. Sorry, I'm not very familiar with MEG, I wasn't aware of this.

File: mne_bids/utils.py [hunk 6]
```diff
@@ -89,22 +89,67 @@ def _get_ch_type_mapping(fro='mne', to='bids'):
...
+        datatypes = list()
+        ieeg_types = ['seeg', 'ecog']
+        if any(ieeg_type in raw for ieeg_type in ieeg_types):
+            datatypes.append('ieeg')
+        if 'meg' in raw:
+            datatypes.append('meg')
+        if 'eeg' in raw:
+            datatypes.append('eeg')
+        if len(datatypes) == 0:
+            raise ValueError('No MEG, EEG or iEEG channels found in data.'
```
Comment (agramfort) [inline mne_bids/utils.py:136]:
```suggestion
            raise ValueError('No MEG, EEG or iEEG channels found in data. '
```

File: mne_bids/utils.py [hunk 7]
```diff
@@ -385,3 +430,42 @@ def _stamp_to_dt(utc_stamp):
...
+            f'It should be one of  either `meg`, `eeg` or `ieeg` (Got '
+            f'`{datatype}`. Please specify a valid datatype using '
+            f'`bids_path.update(datatype="<datatype>")`.')
+    datatype_matches = False
+    if datatype == 'eeg' and datatype in raw:
+        datatype_matches = True
+    elif datatype == 'meg' and datatype in raw:
+        datatype_matches = True
+    elif datatype == 'ieeg':
+        ieeg_types = ['seeg', 'ecog']
```
Comment (agramfort) [inline mne_bids/utils.py:462]:
```suggestion
        ieeg_types = ('seeg', 'ecog')
```

File: mne_bids/utils.py [hunk 8]
```diff
@@ -385,3 +430,42 @@ def _stamp_to_dt(utc_stamp):
...
+    raw : mne.io.Raw
+        Raw object.
+    datatype : str
+        Can be one of either ``'meg'``, ``'eeg'``, or ``'ieeg'``.
+
+    Returns
+    -------
+    None
+    """
+    supported_types = ['meg', 'eeg', 'ieeg']
```
Comment (agramfort) [inline mne_bids/utils.py:449]:
```suggestion
    supported_types = ('meg', 'eeg', 'ieeg')
```

File: mne_bids/tests/test_utils.py [hunk 9]
```diff
@@ -124,3 +144,47 @@ def test_infer_eeg_placement_scheme():
...
+        _check_datatype(raw_eeg, datatype)
+    datatype = 'anat'
+    with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
+                                         'is currently not supported.'):
+        _check_datatype(raw_eeg, datatype)
+    datatype = 'eeg'
+    _check_datatype(raw_eeg, datatype)
+    with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
+                                         'was not found in the raw object.'):
+        _check_datatype(raw_meg, datatype)
```
Comment (agramfort) [inline mne_bids/tests/test_utils.py:171]:
```suggestion
    for datatype, raw in [(None, raw_eeg, ('eeg', raw_meg), ('anat', raw_eeg)]:
        with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
                                             'is currently not supported.'):
            _check_datatype(raw, datatype)
```

can you compress with something like this?

Comment (richardkoehler) [reply mne_bids/tests/test_utils.py]: I did, hope it is fine the way I have solved it now :)

**Review by agramfort — APPROVED** [review]
thx @richardkoehler !

**Comment by agramfort** [comment]
thx a lot @richardkoehler !

**Comment by richardkoehler** [comment]
@agramfort  Thanks for the merge, happy to contribute :) Stupid question because this is my first contribute to mne-bids: The Merge checklist at the top is not for myself to complete right? Because it says "This is not your own PR". I haven't updated whats_new.rst or anything, but of course I would be happy to do so, if required

**Comment by agramfort** [comment]
it's for you to fill to help us review but here I looked myself.

thx again

>

**Comment by hoechenberger** [comment]
Thanks @richardkoehler!

**Comment by hoechenberger** [comment]
Oh noes, we forgot to add a changelog entry!! @richardkoehler care to add one?

**Comment by richardkoehler** [comment]
@hoechenberger Yes, I did so in #779 as I couldn't reopen this PR.

**Pull request merged**